<!DOCTYPE html>
<html lang="en-us">
  <head>
    
    <script type="application/ld+json">

{  
  "@context":"http://schema.org",
  "@type":"Website",
  "@id":"\/",
  "author": {
    "@type": "Person",
    "name": "Christoph Freier",
    
    "image": "/image/bio.png"
    
  },
  "name":"In Code We Trust",
  "description":"What if there is no prediction: the challenge of unsupervised learning.",
  "url":"\/2019\/11\/visualization-and-clustering-for-single-cell-analysis\/",
  "keywords":"[, rstats, data science]"
}

</script>
    <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="generator" content="Hugo 0.120.4 with theme Tranquilpeak 0.5.3-BETA">
<meta name="author" content="Christoph Freier">
<meta name="keywords" content=", rstats, data science">
<meta name="description" content="What if there is no prediction: the challenge of unsupervised learning.">


<meta property="og:description" content="What if there is no prediction: the challenge of unsupervised learning.">
<meta property="og:type" content="article">
<meta property="og:title" content="Visualization and Clustering for Single-Cell Analysis">
<meta name="twitter:title" content="Visualization and Clustering for Single-Cell Analysis">
<meta property="og:url" content="/2019/11/visualization-and-clustering-for-single-cell-analysis/">
<meta property="twitter:url" content="/2019/11/visualization-and-clustering-for-single-cell-analysis/">
<meta property="og:site_name" content="In Code We Trust">
<meta property="og:description" content="What if there is no prediction: the challenge of unsupervised learning.">
<meta name="twitter:description" content="What if there is no prediction: the challenge of unsupervised learning.">
<meta property="og:locale" content="en-us">

  
    <meta property="article:published_time" content="2019-11-27T00:00:00">
  
  
    <meta property="article:modified_time" content="2019-11-27T00:00:00">
  
  
  
    
      <meta property="article:section" content="Unsupervised Learning">
    
      <meta property="article:section" content="Clustering">
    
  
  
    
      <meta property="article:tag" content="tidyverse">
    
      <meta property="article:tag" content="RColorBrewer">
    
      <meta property="article:tag" content="Rtsne">
    
      <meta property="article:tag" content="umap">
    
      <meta property="article:tag" content="Rphenograph">
    
      <meta property="article:tag" content="viridis">
    
      <meta property="article:tag" content="genefilter">
    
  


<meta name="twitter:card" content="summary">

  <meta name="twitter:site" content="@ChristophFreier">


  <meta name="twitter:creator" content="@ChristophFreier">






  <meta property="og:image" content="/image/bio.png">
  <meta property="twitter:image" content="/image/bio.png">




  <meta property="og:image" content="https://raw.githubusercontent.com/Kawameicha/Kawameicha.github.io/master/image/cover/cell.png">
  <meta property="twitter:image" content="https://raw.githubusercontent.com/Kawameicha/Kawameicha.github.io/master/image/cover/cell.png">


  <meta property="og:image" content="https://raw.githubusercontent.com/Kawameicha/Kawameicha.github.io/master/image/thumbnail/disease.png">
  <meta property="twitter:image" content="https://raw.githubusercontent.com/Kawameicha/Kawameicha.github.io/master/image/thumbnail/disease.png">


    <title>Visualization and Clustering for Single-Cell Analysis</title>

    <link rel="icon" href="../../../image/favicon.ico">
    

    

    <link rel="canonical" href="../../../2019/11/visualization-and-clustering-for-single-cell-analysis/">

    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/all.min.css" integrity="sha512-iBBXm8fW90+nuLcSKlbmrPcLa0OT92xO1BIsZ+ywDWZCvqsWgccV3gFoRBv0z+8dLJgyAHIhR35VZc2oM/gI1w==" crossorigin="anonymous" referrerpolicy="no-referrer" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.css" integrity="sha512-H9jrZiiopUdsLpg94A333EfumgUBpO9MdbxStdeITo+KEIMaNfHNvwyjjDJb+ERPaRS6DpyRlKbvPUasNItRyw==" crossorigin="anonymous" referrerpolicy="no-referrer" />
    
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/gitalk/1.7.2/gitalk.css" integrity="sha512-MLcK/YRapzET1qTBXrOiZE6bGBgtATMo2bIyalVJ8EKDEGNoeA3SPQkvWAR0zNS650YG13ocXBMeioDuZcSRuQ==" crossorigin="anonymous" referrerpolicy="no-referrer" />
    
    
    
    <link rel="stylesheet" href="../../../css/style-h6ccsoet3mzkbb0wngshlfbaweimexgqcxj0h5hu4h82olsdzz6wmqdkajm.min.css" />
    
    

    
      
<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
	(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
	m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
	})(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
	ga('create', 'UA-166670924-1', 'auto');
	
	ga('send', 'pageview');
}
</script>
    
    
  </head>

  <body>
    <div id="blog">
      <header id="header" data-behavior="5">
  <i id="btn-open-sidebar" class="fa fa-lg fa-bars"></i>
  <div class="header-title">
    <a class="header-title-link" href="../../../" aria-label="Go to homepage">In Code We Trust</a>
  </div>
  
    
      <a class="header-right-picture "
         href="../../../#about" aria-label="Open the link: /#about">
    
    
    
      
        <img class="header-picture" src="../../../image/bio.png" alt="Author&#39;s picture" />
      
    
    </a>
  
</header>

      <nav id="sidebar" data-behavior="5">
  <div class="sidebar-container">
    
      <div class="sidebar-profile">
        <a href="../../../#about" aria-label="Read more about the author">
          <img class="sidebar-profile-picture" src="../../../image/bio.png" alt="Author&#39;s picture" />
        </a>
        <h4 class="sidebar-profile-name">Christoph Freier</h4>
        
          <h5 class="sidebar-profile-bio">Hello, my name is Christoph Freier. Regularly, people stop me on the street and ask me <em>&ldquo;hey, what&rsquo;s your name?&rdquo;</em> I reply my name is Christoph Freier and I rock a great mustache.</h5>
        
      </div>
    
    <ul class="sidebar-buttons">
      
  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="../../../" title="Home">
    
      <i class="sidebar-button-icon fa fa-lg fa-home"></i>
      
      <span class="sidebar-button-desc">Home</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="../../../categories" title="Categories">
    
      <i class="sidebar-button-icon fa fa-lg fa-sitemap"></i>
      
      <span class="sidebar-button-desc">Categories</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="../../../tags" title="Tags">
    
      <i class="sidebar-button-icon fa fa-lg fa-tags"></i>
      
      <span class="sidebar-button-desc">Tags</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="../../../page/projects" title="Projects">
    
      <i class="sidebar-button-icon fa fa-lg fa-archive"></i>
      
      <span class="sidebar-button-desc">Projects</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="../../../page/useful-resources" title="Resources">
    
      <i class="sidebar-button-icon fa fa-lg fa-book"></i>
      
      <span class="sidebar-button-desc">Resources</span>
    </a>
  </li>


    </ul>
    <ul class="sidebar-buttons">
      
  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="../../../page/about-me" title="About Me">
    
      <i class="sidebar-button-icon fas fa-lg fa-id-card" aria-hidden="true"></i>
      
      <span class="sidebar-button-desc">About Me</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="https://github.com/kawameicha" title="GitHub">
    
      <i class="sidebar-button-icon fab fa-lg fa-github" aria-hidden="true"></i>
      
      <span class="sidebar-button-desc">GitHub</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="https://www.linkedin.com/in/christophfreier/" title="LinkedIn">
    
      <i class="sidebar-button-icon fab fa-lg fa-linkedin" aria-hidden="true"></i>
      
      <span class="sidebar-button-desc">LinkedIn</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="https://stackoverflow.com/users/10705596/christoph-freier" title="Stack Overflow">
    
      <i class="sidebar-button-icon fab fa-lg fa-stack-overflow" aria-hidden="true"></i>
      
      <span class="sidebar-button-desc">Stack Overflow</span>
    </a>
  </li>

  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="https://twitter.com/ChristophFreier" title="Twitter">
    
      <i class="sidebar-button-icon fab fa-lg fa-twitter" aria-hidden="true"></i>
      
      <span class="sidebar-button-desc">Twitter</span>
    </a>
  </li>


    </ul>
    <ul class="sidebar-buttons">
      
  <li class="sidebar-button">
    
      <a class="sidebar-button-link " href="mailto:christoph@freier.fr" target="_blank" rel="noopener" title="Email me">
    
      <i class="sidebar-button-icon fa fa-lg fa-envelope"></i>
      
      <span class="sidebar-button-desc">Email me</span>
    </a>
  </li>


    </ul>
  </div>
</nav>

      
  <div class="post-header-cover
              text-left
              post-header-cover--partial"
       style="background-image:url('https://raw.githubusercontent.com/Kawameicha/Kawameicha.github.io/master/image/cover/cell.png')"
       data-behavior="5">
    
      <div class="post-header main-content-wrap text-left">
  
    <h1 class="post-title">
      Visualization and Clustering for Single-Cell Analysis
    </h1>
  
  
  <div class="postShorten-meta post-meta">
    
      <time datetime="2019-11-27T00:00:00Z">
        
  
  
  
  
    Nov 27, 2019
  

      </time>
    
    
  
  
    <span>in</span>
    
      <a class="category-link" href="../../../categories/unsupervised-learning">Unsupervised Learning</a>, 
    
      <a class="category-link" href="../../../categories/clustering">Clustering</a>
    
  

  </div>

</div>
    
  </div>


      <div id="main" data-behavior="5"
        class="hasCover
               hasCoverMetaIn
               ">
        <article class="post" id="top">
          
          
          <div class="post-content markdown">
            <div class="main-content-wrap">
              
<script src="../../../rmarkdown-libs/header-attrs/header-attrs.js"></script>


<p>Let’s talk biology. That’s my background and it has become really hype lately with several companies now offering to sequence your genome paving the way to better understanding of the molecular basis of disease and personalized medicine. Nowadays, there is a dramatic increase in the number of large, highly complex datasets being generated from experiments, quantifying as many different variables such as gene, protein, or population-wide genetic variation. Here, there is no interest in prediction, mostly because there is no associated response variable.</p>
<p>Consider single cells. It’s much easier–as well as way cheaper–to sequence a bunch of single cells rather than obtaining the same amount of information about single individuals. Unfortunately, the budget to write these blog posts is limited by the amount of out of my pocket money I can inject in and this quite directly correlates with my pocket’s size. Long story short, there is no way I can compete with big pharmaceutical companies. I found these 301 cells open access on the web (<a href="https://www.nature.com/articles/nbt.2967">Pollen <em>et al.</em>, 2014</a>).</p>
<p>After some initial cleaning, 100 genes of interest have been selected that we can use to discover patterns. In this case the genes are the observations and the patterns could be different cell populations.</p>
<div id="why-wouldnt-we-simply-compute-all-possibilities" class="section level1">
<h1>Why wouldn’t we simply compute all possibilities</h1>
<p>Good question! Glad you asked, really. This certainly has a lot to do with the number of hypothesis growing exponentially with the number of attributes. If we define the attributes here as Boolean, that is cell <code>c</code> expresses gene <code>g</code> yes or no, we clearly oversimplify this problem. Nonetheless, each new attribute doubles the number of possible instances by taking each previous instance and extending it with a yes or no for that attribute. In turn, the rate of change–the change per instant or unit of time–becomes proportional to the current value.</p>
<p>Think about a tiny <a href="https://en.m.wikipedia.org/wiki/Escherichia_coli"><em>Escherichia coli</em></a>, commonly found in the lower intestinal tract of warm blooded organisms and that will happily divide into two every 15 minutes (roughly) provided there is enough nutrient.</p>
<pre class="r"><code># Define an expGro function
expGro = function(x){(1e-18)*(2^x)}

# Feed the E. coli
ggplot(data.frame(x = c(1, 130)), aes(x = x)) +
  stat_function(fun = expGro, geom = &quot;line&quot; ) +
  theme_minimal() +
  scale_x_continuous(breaks = c(40, 80, 120),
                     labels = c(&quot;10&quot;, &quot;20&quot;, &quot;30&quot;)) +
  geom_hline(aes(yintercept = 1.09751e+21), colour = &quot;#EB821B&quot;, linetype = &quot;dashed&quot;) +
  annotate(&quot;text&quot;, x = 10, y = 1.14875e+21, colour = &quot;#EB821B&quot;, label = &quot;Volume of the Earth&quot;) +
  labs(x = &quot;Hours&quot;, y = &quot;Volume&quot;,
       title    = &quot;Exponential Growth is Scary&quot;,
       subtitle = &quot;Growth of a single E. coli&quot;)</code></pre>
<p><img src="../../../post/2019-11-27-visualization-and-clustering-for-single-cell-analysis_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
<p>Very scary indeed. It would take about one day and a half starting with a single <em>Escherichia coli</em> to divide into a bacterial mass as big as the Earth itself. Think about this next time you consider adding sugar to your coffee, or worse, succumb to temptation and getting donuts too.</p>
</div>
<div id="the-truth-is-generally-seen-rarely-heard-b.-gracian" class="section level1">
<h1>The truth is generally seen, rarely heard (B. Gracian)</h1>
<p>Unsupervised learning is more descriptive and definitively more subjective than its supervised counterpart as there is neither simple goal nor universally accepted cross validation. It should be approached as part of an exploratory data analysis where visualization is fundamental. A good visualization will show you things that you didn’t expect, or help raise new questions about the data.</p>
<p>We are here faced with a large set of correlated variables, <em>i.e.</em> gene expression. Each cell might have a unique signature but cells belonging to the same population will tend to have a similar profile and principal components allow us to summarize this dataset with a smaller number of representative variables that collectively should explain most of the variability. It is a powerful approach that usually performs really well on this type of data even though it is quite naive to think all relations could be linear.</p>
<pre class="r"><code># Compute principal components
pca &lt;- pollen %&gt;% 
  select_if(is.numeric) %&gt;% 
  prcomp()

pca_plot &lt;- data.frame(x = pca$x[,1],
                       y = pca$x[,2])</code></pre>
<pre class="r"><code>ggplot(pca_plot) +
  geom_point(aes(x = x, y = y)) + 
  theme_minimal() +
  theme(legend.title = element_blank(),
        axis.text    = element_blank(),
        axis.title   = element_blank())</code></pre>
<p><img src="../../../post/2019-11-27-visualization-and-clustering-for-single-cell-analysis_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<p>There you go, populations! No surprise here, it already works really well. But are there alternatives?</p>
<p>Yes, looking at nonlinear dimensionality reduction method, t-SNE automatically pops into my mind. It can deal with more complex nonlinear patterns of Gaussian clusters in multidimensional space compared to PCA so it is good for single cell RNA-sequencing analysis. Although is not suited to finding outliers because of the way the samples are arranged that does not directly represent distance, like in PCA. Instead t-SNE preserves local distance at the expensive of global distance. And it’s easy to run in R.</p>
<pre class="r"><code># Run t-SNE
tsne &lt;- pollen %&gt;% 
  select_if(is.numeric) %&gt;% 
  Rtsne()

tsne_plot &lt;- data.frame(x = tsne$Y[,1],
                        y = tsne$Y[,2])</code></pre>
<pre class="r"><code>ggplot(tsne_plot) +
  geom_point(aes(x = x, y = y)) + 
  theme_minimal() +
  theme(legend.title = element_blank(),
        axis.text    = element_blank(),
        axis.title   = element_blank())</code></pre>
<p><img src="../../../post/2019-11-27-visualization-and-clustering-for-single-cell-analysis_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>Neat! By the way, a PCA was used here too as a preprocessing step to get rid of all the zeros.</p>
<p>Oh yes, and of course there is UMAP. UMAP is a non linear dimensionality reduction algorithm of the same family as t-SNE. In the first phase of UMAP a weighted <em>k</em>-nearest neighbor graph is computed, in the second a low dimensionality layout of this is then calculated. Then the embedded data points can be visualized in a new space and compared with other variables of interest. It can be used for the analysis of many types of data, including, single-cell RNA-sequencing and better preserves global distances.</p>
<pre class="r"><code># Perform UMAP
umap &lt;- pollen %&gt;% 
  select_if(is.numeric) %&gt;% 
  umap()

umap_plot &lt;- data.frame(x = umap$layout[,1],
                        y = umap$layout[,2])</code></pre>
<pre class="r"><code>ggplot(umap_plot) +
  geom_point(aes(x = x, y = y)) + 
  theme_minimal() +
  theme(legend.title = element_blank(),
        axis.text    = element_blank(),
        axis.title   = element_blank())</code></pre>
<p><img src="../../../post/2019-11-27-visualization-and-clustering-for-single-cell-analysis_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>Yes, I definitively like this one a lot! You can add a color scale and risk an educated guess.</p>
<pre class="r"><code>ggplot(umap_plot) +
  geom_point(aes(x = x, y = y, col = pollen$HBG2)) + 
  scale_colour_gradient(low  = &quot;#8acf81&quot;,
                        high = &quot;#ef5534&quot;,
                        name = &quot;HBG2&quot;) +
  theme_minimal() +
  theme(axis.text  = element_blank(),
        axis.title = element_blank())</code></pre>
<p><img src="../../../post/2019-11-27-visualization-and-clustering-for-single-cell-analysis_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
<p>You’ve probably checked already, but <code>prcomp</code> is part of the <code>stats</code> packages, while <code>Rtsne</code> and <code>umap</code> are part of the, well, <code>Rtsne</code> and <code>umap</code> libraries, respectively. There are many arguments you can play with to adapt this to other datasets and get better embedding in a lower number of dimensions in order to see patterns and trends in the data.</p>
</div>
<div id="algorithms-are-just-opinions-expressed-in-numbers-c.-oneil" class="section level1">
<h1>Algorithms are just opinions expressed in numbers (C. O’Neil)</h1>
<p>As good as visualization is, it doesn’t scale up particularly well. Once we have observed these populations, we certainly would like to cluster them. That is partition them into distinct groups so that the observations within each group are quite similar to each other, while observations in different groups are quite different from each other.</p>
<p>One of the simplest, but still very elegant approach to perform this is the <em>k</em>-means clustering that will partition a dataset into <em>K</em> distinct, non-overlapping clusters. Like all clustering methods, it is biased and not completely arbitrary. We must first specify the desired number of cluster <em>K</em>. As a rule of thumb, I suggest we use our previous observations of about 9 islands and add ~10%. It is well known that some people have a green thumb. Mine is normally pink but gives me this unusual ability to find a good <em>K</em>.</p>
<pre class="r"><code># Set seed
set.seed(123)</code></pre>
<pre class="r"><code># Cluster dataset
kmeans &lt;- pollen %&gt;% 
  select_if(is.numeric) %&gt;% 
  kmeans(10, nstart = 20)</code></pre>
<pre class="r"><code>ggplot(umap_plot) +
  geom_point(aes(x = x, y = y, col = as.factor(kmeans$cluster))) +
  scale_color_brewer(palette = &quot;Set3&quot;) +
  theme_minimal() +
  theme(legend.position = &quot;none&quot;,
        axis.text       = element_blank(),
        axis.title      = element_blank())</code></pre>
<p><img src="../../../post/2019-11-27-visualization-and-clustering-for-single-cell-analysis_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<p>That looks about right. As I said simple but still very elegant. What else?</p>
<p>Hierarchical clustering is an alternative approach which does not force us to commit to a particular number choice of <em>K</em>. At least not upfront. It results in a dendrogram that can be used to obtain any number of clusters. However, often the choice of where to cut isn’t really clear and a visual inspection only slightly helps to select a sensible number of clusters, based on the height of the fusion. We will skip this step here, as 10 looked like a good approximation.</p>
<pre class="r"><code># Cluster dataset
hClust &lt;- pollen %&gt;% 
  select_if(is.numeric) %&gt;% 
  dist() %&gt;% 
  hclust() %&gt;% 
  cutree(10)</code></pre>
<pre class="r"><code>ggplot(umap_plot) +
  geom_point(aes(x = x, y = y, col = as.factor(hClust))) +
  scale_color_brewer(palette = &quot;Set3&quot;) +
  theme_minimal() +
  theme(legend.position = &quot;none&quot;,
        axis.text       = element_blank(),
        axis.title      = element_blank())</code></pre>
<p><img src="../../../post/2019-11-27-visualization-and-clustering-for-single-cell-analysis_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>Looks okay too. Although again, not perfect. Are there any other options?</p>
<p>Alternatively, we could use the Phenograph algorithm, which is a clustering method designed more specifically for high-dimensional single-cell data analysis. It works by creating a graph (“network”) representing phenotypic similarities between cells by calculating the Jaccard coefficient between nearest-neighbor sets, and then identifying communities using the well known Louvain method in this graph. In this case the <em>k</em> is the number of nearest neighbors considered by the algorithm.</p>
<pre class="r"><code># Cluster dataset
pheno &lt;- pollen %&gt;% 
  select_if(is.numeric) %&gt;% 
  Rphenograph(k = 15)</code></pre>
<pre class="r"><code>ggplot(umap_plot) +
  geom_point(aes(x = x, y = y, col = as.factor(membership(pheno[[2]])))) +
  scale_color_brewer(palette = &quot;Set3&quot;) +
  theme_minimal() +
  theme(legend.position = &quot;none&quot;,
        axis.text       = element_blank(),
        axis.title      = element_blank())</code></pre>
<p><img src="../../../post/2019-11-27-visualization-and-clustering-for-single-cell-analysis_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<p>Result is clearly improved here and we seem to have captured all islands.</p>
<p>As previously mentioned, unsupervised learning is rather subjective and it usually isn’t possible to access the quality of the results obtained. However, in this particular case, the dataset indicates the real cell type. This looks pretty close to the Phenograph algorithm result.</p>
<pre class="r"><code>ggplot(umap_plot) +
  geom_point(aes(x = x, y = y, col = pollen$pollen.celltypes)) +
  scale_color_brewer(palette = &quot;Set3&quot;) +
  theme_minimal() +
  theme(legend.title = element_blank(),
        axis.text    = element_blank(),
        axis.title   = element_blank())</code></pre>
<p><img src="../../../post/2019-11-27-visualization-and-clustering-for-single-cell-analysis_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
<p>The only populations we were not able to capture correctly are <code>GW16</code>, <code>GW21</code>, and <code>GW21+3</code>. This isn’t really surprising considering these appear to be very close on the visualization and as one can guess from their names, are very similar indeed. Although treated, they don’t differ from each other.</p>
</div>
<div id="alright-can-we-compare-these-populations-now" class="section level1">
<h1>Alright, can we compare these populations now</h1>
<p>I believe we do yes.</p>
<p>Simply pivoting the data using the new <code>pivot_longer</code> from <code>tidyr</code> we can visualize as beautiful heatmap with a little help from the <code>viridis</code> package. Here for the four first clusters only.</p>
<pre class="r"><code># Pivot the data
pheno_tbl &lt;- pollen %&gt;% 
  select_if(is.numeric) %&gt;%
  add_column(as.factor(membership(pheno[[2]]))) %&gt;% 
  rename(cluster = `as.factor(membership(pheno[[2]]))`) %&gt;% 
  pivot_longer(-cluster, names_to = &quot;gene&quot;, values_to = &quot;expression&quot;)</code></pre>
<pre class="r"><code>pheno_tbl %&gt;% 
  filter(cluster %in% c(1:4)) %&gt;% 
  ggplot(aes(x = gene, y = cluster, fill = log2(expression+.1))) +
  geom_tile(color = &quot;white&quot;, size = .1) +
  scale_fill_viridis() +
  theme_minimal() +
  coord_flip() +
  theme(legend.position = &quot;none&quot;,
        axis.text.y     = element_text(size = 4),
        axis.title      = element_blank())</code></pre>
<p><img src="../../../post/2019-11-27-visualization-and-clustering-for-single-cell-analysis_files/figure-html/unnamed-chunk-19-1.png" width="672" /></p>
<p>It’s a bit messy, agreed. Let’s filter the most variable genes using the <code>genefilter</code> library to visualize them as violin plot. Here the four most variable ones only.</p>
<pre class="r"><code># Select four most variable genes
topVar &lt;- pollen %&gt;% 
  select(-contains(&quot;Spike&quot;)) %&gt;% 
  select_if(is.numeric) %&gt;%
  t %&gt;% 
  rowVars %&gt;% 
  head(4)</code></pre>
<pre class="r"><code>pheno_tbl %&gt;% 
  filter(gene %in% names(topVar)) %&gt;% 
  ggplot(aes(x = cluster, y = expression, fill = gene)) +
  geom_violin(scale = &quot;width&quot;) +
  scale_fill_brewer(palette = &quot;Set3&quot;) +
  theme_minimal() +
  theme(legend.position = &quot;none&quot;,
        axis.text.y     = element_blank(),
        axis.title      = element_blank()) +
  facet_wrap(~gene)</code></pre>
<p><img src="../../../post/2019-11-27-visualization-and-clustering-for-single-cell-analysis_files/figure-html/unnamed-chunk-21-1.png" width="672" /></p>
</div>
<div id="subject-expertise-matters" class="section level1">
<h1>Subject expertise matters</h1>
<p>While doing unsupervised learning, instead of predicting a particular output variable, we are interested in determining whether there are groups, or clusters, among the observations based on a set of measurements. In this type of learning we seek to find previously unknown patterns in a dataset without pre-existing labels. This can be applied in many different domains to uncover the existence of rare groups within a larger population that may never have been seen before. Although it was already possible to study the profile of single cells for a couple of markers, single-cell RNA-sequencing now allows to study the expression profiles of individual cells. And that’s exciting!</p>
<p>It is not possible to obtain complete information on every RNA expressed by each cell, due to the small amount of material available, but patterns of gene expression can be identified through gene clustering analyses. In this particular dataset, we have 11 different cell types and we manage to visualize all cell lines. A good visualization is essential and can also hint that you’re asking the wrong question, or you need to collect different data. Once you have made your questions sufficiently precise, you can use a model to answer them.</p>
</div>

              


            </div>
          </div>
          <div id="post-footer" class="post-footer main-content-wrap">
            
              
                
                
                  <div class="post-footer-tags">
                    <span class="text-color-light text-small">TAGGED IN</span><br/>
                    
  <a class="tag tag--primary tag--small" href="../../../tags/tidyverse/">tidyverse</a>

  <a class="tag tag--primary tag--small" href="../../../tags/rcolorbrewer/">RColorBrewer</a>

  <a class="tag tag--primary tag--small" href="../../../tags/rtsne/">Rtsne</a>

  <a class="tag tag--primary tag--small" href="../../../tags/umap/">umap</a>

  <a class="tag tag--primary tag--small" href="../../../tags/rphenograph/">Rphenograph</a>

  <a class="tag tag--primary tag--small" href="../../../tags/viridis/">viridis</a>

  <a class="tag tag--primary tag--small" href="../../../tags/genefilter/">genefilter</a>

                  </div>
                
              
            
            
<div class="post-actions-wrap">
  <nav >
    <ul class="post-actions post-action-nav">
      
        <li class="post-action">
          
            <a class="post-action-btn btn btn--default tooltip--top" href="../../../2020/02/the-king-in-yellow/" data-tooltip="The King in Yellow" aria-label="NEXT: The King in Yellow">
          
              <i class="fa fa-angle-left"></i>
              <span class="hide-xs hide-sm text-small icon-ml">NEXT</span>
            </a>
        </li>
        <li class="post-action">
          
            <a class="post-action-btn btn btn--default tooltip--top" href="../../../2018/11/my-heart-will-go-down/" data-tooltip="My Heart Will Go Down" aria-label="PREVIOUS: My Heart Will Go Down">
          
              <span class="hide-xs hide-sm text-small icon-mr">PREVIOUS</span>
              <i class="fa fa-angle-right"></i>
            </a>
        </li>
      
    </ul>
  </nav>
<ul class="post-actions post-action-share" >
  
    <li class="post-action hide-lg hide-md hide-sm">
      <a class="post-action-btn btn btn--default btn-open-shareoptions" href="#btn-open-shareoptions" aria-label="Share this post">
        <i class="fa fa-share-alt" aria-hidden="true"></i>
      </a>
    </li>
    
      <li class="post-action hide-xs">
        <a class="post-action-btn btn btn--default" target="new" href="https://www.linkedin.com/sharing/share-offsite/?url=../../../2019/11/visualization-and-clustering-for-single-cell-analysis/" title="Share on Linkedin" aria-label="Share on Linkedin">
          <i class="fab fa-linkedin" aria-hidden="true"></i>
        </a>
      </li>
    
      <li class="post-action hide-xs">
        <a class="post-action-btn btn btn--default" target="new" href="https://www.facebook.com/sharer/sharer.php?u=/2019/11/visualization-and-clustering-for-single-cell-analysis/" title="Share on Facebook" aria-label="Share on Facebook">
          <i class="fab fa-facebook-square" aria-hidden="true"></i>
        </a>
      </li>
    
      <li class="post-action hide-xs">
        <a class="post-action-btn btn btn--default" target="new" href="https://twitter.com/intent/tweet?text=/2019/11/visualization-and-clustering-for-single-cell-analysis/" title="Share on Twitter" aria-label="Share on Twitter">
          <i class="fab fa-twitter" aria-hidden="true"></i>
        </a>
      </li>
    
  
  
    <li class="post-action">
      <a class="post-action-btn btn btn--default" href="#gitalk" aria-label="Leave a comment">
        <i class="fa fa-comment"></i>
      </a>
    </li>
  
  <li class="post-action">
    
      <a class="post-action-btn btn btn--default" href="#top" aria-label="Back to top">
      <i class="fa fa-arrow-up" aria-hidden="true"></i>
    
    </a>
  </li>
</ul>
</div>


            
  
    <div id="gitalk">
      <noscript>Please enable JavaScript to view the comments powered by Gitalk.</noscript>
    </div>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/gitalk/1.7.2/gitalk.min.js" integrity="sha512-EcTCcXV46teiNwe0VcnM5A038tcY+BaQYO4nW6Gh2i7v4/HjBVg7xx3+JBLl9WofDds//INJAiEGAtdgr8PWyA==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
    <script type="text/javascript">
      (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
          document.getElementById('gitalk').innerHTML = 'Gitalk comments not available by default when the website is previewed locally.';
          return;
        }
        new Gitalk({
          clientID: '29291d03f06112f6b4c1',
          clientSecret: '99928d6b52d8e52b3da29bc86598692956f21c16',
          repo: 'Gitalk',
          owner: 'Kawameicha',
          admin: ['Kawameicha'],
          id: '50458c33457c39a81b4921e24c2da0a7',
          ...{"distractionfreemode":false,"enablehotkey":true,"language":"en","pagerdirection":"first","perpage":10}
        }).render('gitalk')
      })()
    </script>
  


          </div>
        </article>
        <footer id="footer" class="main-content-wrap">
  <span class="copyrights">
    &copy; 2023 Christoph Freier. All Rights Reserved
  </span>
</footer>

      </div>
      <div id="bottom-bar" class="post-bottom-bar" data-behavior="5">
        
<div class="post-actions-wrap">
  <nav >
    <ul class="post-actions post-action-nav">
      
        <li class="post-action">
          
            <a class="post-action-btn btn btn--default tooltip--top" href="../../../2020/02/the-king-in-yellow/" data-tooltip="The King in Yellow" aria-label="NEXT: The King in Yellow">
          
              <i class="fa fa-angle-left"></i>
              <span class="hide-xs hide-sm text-small icon-ml">NEXT</span>
            </a>
        </li>
        <li class="post-action">
          
            <a class="post-action-btn btn btn--default tooltip--top" href="../../../2018/11/my-heart-will-go-down/" data-tooltip="My Heart Will Go Down" aria-label="PREVIOUS: My Heart Will Go Down">
          
              <span class="hide-xs hide-sm text-small icon-mr">PREVIOUS</span>
              <i class="fa fa-angle-right"></i>
            </a>
        </li>
      
    </ul>
  </nav>
<ul class="post-actions post-action-share" >
  
    <li class="post-action hide-lg hide-md hide-sm">
      <a class="post-action-btn btn btn--default btn-open-shareoptions" href="#btn-open-shareoptions" aria-label="Share this post">
        <i class="fa fa-share-alt" aria-hidden="true"></i>
      </a>
    </li>
    
      <li class="post-action hide-xs">
        <a class="post-action-btn btn btn--default" target="new" href="https://www.linkedin.com/sharing/share-offsite/?url=../../../2019/11/visualization-and-clustering-for-single-cell-analysis/" title="Share on Linkedin" aria-label="Share on Linkedin">
          <i class="fab fa-linkedin" aria-hidden="true"></i>
        </a>
      </li>
    
      <li class="post-action hide-xs">
        <a class="post-action-btn btn btn--default" target="new" href="https://www.facebook.com/sharer/sharer.php?u=/2019/11/visualization-and-clustering-for-single-cell-analysis/" title="Share on Facebook" aria-label="Share on Facebook">
          <i class="fab fa-facebook-square" aria-hidden="true"></i>
        </a>
      </li>
    
      <li class="post-action hide-xs">
        <a class="post-action-btn btn btn--default" target="new" href="https://twitter.com/intent/tweet?text=/2019/11/visualization-and-clustering-for-single-cell-analysis/" title="Share on Twitter" aria-label="Share on Twitter">
          <i class="fab fa-twitter" aria-hidden="true"></i>
        </a>
      </li>
    
  
  
    <li class="post-action">
      <a class="post-action-btn btn btn--default" href="#gitalk" aria-label="Leave a comment">
        <i class="fa fa-comment"></i>
      </a>
    </li>
  
  <li class="post-action">
    
      <a class="post-action-btn btn btn--default" href="#top" aria-label="Back to top">
      <i class="fa fa-arrow-up" aria-hidden="true"></i>
    
    </a>
  </li>
</ul>
</div>


      </div>
      
<div id="share-options-bar" class="share-options-bar" data-behavior="5">
  <i id="btn-close-shareoptions" class="fa fa-times"></i>
  <ul class="share-options">
    
      <li class="share-option">
        <a class="share-option-btn" target="new" href="https://www.linkedin.com/sharing/share-offsite/?url=%2F2019%2F11%2Fvisualization-and-clustering-for-single-cell-analysis%2F" aria-label="Share on Linkedin">
          <i class="fab fa-linkedin" aria-hidden="true"></i><span>Share on Linkedin</span>
        </a>
      </li>
    
      <li class="share-option">
        <a class="share-option-btn" target="new" href="https://www.facebook.com/sharer/sharer.php?u=%2F2019%2F11%2Fvisualization-and-clustering-for-single-cell-analysis%2F" aria-label="Share on Facebook">
          <i class="fab fa-facebook-square" aria-hidden="true"></i><span>Share on Facebook</span>
        </a>
      </li>
    
      <li class="share-option">
        <a class="share-option-btn" target="new" href="https://twitter.com/intent/tweet?text=%2F2019%2F11%2Fvisualization-and-clustering-for-single-cell-analysis%2F" aria-label="Share on Twitter">
          <i class="fab fa-twitter" aria-hidden="true"></i><span>Share on Twitter</span>
        </a>
      </li>
    
  </ul>
</div>
<div id="share-options-mask" class="share-options-mask"></div>


    </div>
    
    <div id="about">
  <div id="about-card">
    <div id="about-btn-close">
      <i class="fa fa-times"></i>
    </div>
    
      <img id="about-card-picture" src="../../../image/bio.png" alt="Author&#39;s picture" />
    
    <h4 id="about-card-name">Christoph Freier</h4>
    
      <div id="about-card-bio">Hello, my name is Christoph Freier. Regularly, people stop me on the street and ask me <em>&ldquo;hey, what&rsquo;s your name?&rdquo;</em> I reply my name is Christoph Freier and I rock a great mustache.</div>
    
    
      <div id="about-card-job">
        <i class="fa fa-briefcase"></i>
        <br/>
        SaaS Application Consulting
      </div>
    
    
      <div id="about-card-location">
        <i class="fa fa-map-marker-alt"></i>
        <br/>
        I live in Berlin, Germany
      </div>
    
  </div>
</div>

    

    
  
    
      <div id="cover" style="background-image:url('/image/cover.jpg');"></div>
    
  


    
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.6.0/jquery.min.js" integrity="sha512-894YE6QWD5I59HgZOGReFYm4dnWc1Qt5NtvYSaNcOP+u1T9qYdvdihz0PPSiiqn/+/3e7Jo4EaG7TubfWGUrMQ==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.5.7/jquery.fancybox.min.js" integrity="sha512-uURl+ZXMBrF4AwGaWmEetzrd+J5/8NRkWAvJx5sbPSSuOb0bZLqf+tOzniObO00BjHa/dD7gub9oCGMLPQHtQA==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>


<script src="../../../js/script-yqzy9wdlzix4lbbwdnzvwx3egsne77earqmn73v9uno8aupuph8wfguccut.min.js"></script>






    
  </body>
</html>

